File: python/number_of_zero_filled_subarrays.py
class Solution:
    def zeroFilledSubarray(self, nums: list[int]) -> int:
        comb: int = 0
        count: int = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                count += 1
            else:
                comb += self.cal_helper(count)
                count = 0

        comb += self.cal_helper(count)

        return comb

    def cal_helper(self, count: int) -> int:
        result: int = 0

        for i in range(1, count + 1):
            result += i

        return result


def test_zeroFilledSubarray_case_1():
    # arrange
    nums: list[int] = [1, 3, 0, 0, 2, 0, 0, 4]
    expected: int = 6

    # act
    solution = Solution()
    actual = solution.zeroFilledSubarray(nums)

    # assert
    assert expected == actual


def test_zeroFilledSubarray_case_2():
    # arrange
    nums: list[int] = [0, 0, 0, 2, 0, 0]
    expected: int = 9

    # act
    solution = Solution()
    actual = solution.zeroFilledSubarray(nums)

    # assert
    assert expected == actual

File: rust/src/code/number_of_zero_filled_subarrays.rs
struct Solution {}

impl Solution {
    // [1, 3, 0, 0, 2, 0, 0, 4]
    // [0, 0]
    // [0, 0, 0, 2, 0, 0]
    pub fn zero_filled_subarray(nums: Vec<i32>) -> i64 {
        let mut count: i64 = 0;
        let mut comb: i64 = 0;
        for i in 0..nums.len() {
            if nums[i] == 0 {
                count += 1;
            } else {
                comb += Self::cal_helper(count);
                count = 0
            }
        }

        comb += Self::cal_helper(count);

        comb
    }

    fn cal_helper(num: i64) -> i64 {
        let mut result = 0;

        if num == 0 {
            return 0;
        }

        for i in 1..num + 1 {
            result += i;
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_filled_subarray_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 3, 0, 0, 2, 0, 0, 4];
        let expected: i64 = 6;

        // act
        let actual = Solution::zero_filled_subarray(nums);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn zero_filled_subarray_case_2() {
        // arrange
        let nums: Vec<i32> = vec![0, 0, 0, 2, 0, 0];
        let expected: i64 = 9;

        // act
        let actual = Solution::zero_filled_subarray(nums);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn zero_filled_subarray_case_3() {
        // arrange
        let nums: Vec<i32> = vec![0, 0];
        let expected: i64 = 3;

        // act
        let actual = Solution::zero_filled_subarray(nums);

        // assert
        assert_eq!(expected, actual);
    }
}
